tip-distributor/src/claim_mev_workflow.rs

```diff
@@ -30,123 +36,408 @@ pub enum ClaimMevError {
+    },
 }
 
-pub fn claim_mev_tips(
-    merkle_root_path: &PathBuf,
-    rpc_url: &str,
+pub async fn claim_mev_tips(
+    merkle_trees: GeneratedMerkleTreeCollection,
+    rpc_url: String,
+    rpc_send_connection_count: u64,
+    max_concurrent_rpc_get_reqs: usize,
     tip_distribution_program_id: &Pubkey,
-    keypair_path: &PathBuf,
+    keypair: Arc<Keypair>,
+    max_loop_retries: u64,
+    max_loop_duration: Duration,
 ) -> Result<(), ClaimMevError> {
-    const MAX_RETRY_DURATION: Duration = Duration::from_secs(600);
+    let payer_pubkey = keypair.pubkey();
+    let blockhash_rpc_client = Arc::new(RpcClient::new_with_commitment(
+        rpc_url.clone(),
+        CommitmentConfig::finalized(),
+    ));
+    let rpc_clients = Arc::new(
+        (0..rpc_send_connection_count)
+            .map(|_| {
+                Arc::new(RpcClient::new_with_commitment(
+                    rpc_url.clone(),
+                    CommitmentConfig::confirmed(),
+                ))
+            })
+            .collect_vec(),
+    );
 
-    let merkle_trees: GeneratedMerkleTreeCollection =
-        read_json_from_file(merkle_root_path).expect("read GeneratedMerkleTreeCollection");
-    let keypair = read_keypair_file(keypair_path).expect("read keypair file");
+    let tree_nodes = merkle_trees
+        .generated_merkle_trees
+        .iter()
+        .flat_map(|tree| &tree.tree_nodes)
+        .collect_vec();
+    let minimum_rent = blockhash_rpc_client
+        .get_minimum_balance_for_rent_exemption(0)
+        .await
+        .expect("Failed to calculate min rent");
 
-    let tip_distribution_config =
-        Pubkey::find_program_address(&[Config::SEED], tip_distribution_program_id).0;
+    // fetch all accounts up front
+    info!(
+        "Starting to fetch accounts for epoch {}",
+        merkle_trees.epoch
+    );
+    let tdas = crate::get_batched_accounts(
+        &blockhash_rpc_client,
+        max_concurrent_rpc_get_reqs,
+        merkle_trees
+            .generated_merkle_trees
+            .iter()
+            .map(|tree| tree.tip_distribution_account)
+            .collect_vec(),
+    )
+    .await
+    .map_err(ClaimMevError::MaxFetchRetriesExceeded)?
+    .into_iter()
+    .filter_map(|(pubkey, maybe_account)| {
+        let Some(account) = maybe_account else {
+            datapoint_warn!(
+                "claim_mev_workflow-account_error",
+                ("epoch", merkle_trees.epoch, i64),
+                ("pubkey", pubkey.to_string(), String),
+                ("account_type", "tip_distribution_account", String),
+                ("error", 1, i64),
+                ("err_type", "fetch", String),
+                ("err_str", "Failed to fetch TipDistributionAccount", String)
+            );
+            return None;
+        };
+
+        let account = match TipDistributionAccount::try_deserialize(&mut account.data.as_slice()) {
+            Ok(a) => a,
+            Err(e) => {
+                datapoint_warn!(
```

Comment:
would be nice to get a dashboard on these metrics

Reply:
made a ticket: https://linear.app/jito/issue/JIT-1763/optimize-inbound-metrics